asana/asana.go

```diff
@@ -28,8 +28,18 @@ var defaultOptFields = map[string][]string{
 }
 
 type (
+	//Doer interface used for doing http calls.
+	//Use it as point of setting Auth header or custom status code error handling.
+	Doer interface {
+		Do(req *http.Request) (*http.Response, error)
+	}
```

Comment:
Is this client fully taking on responsibility for authentication, or only for certain calls, like in your c.Issue method?  go-github's Client knows nothing about authentication at all.  It is always punted to the underlying http.Client, which typically comes from golang.org/x/oauth2, or the basic auth client that go-github does provide (but is otherwise completely separated from the Client struct).

I think the best analogy to this in go-github would be the custom accept headers we set on various API calls.  That happens by first building the http.Request, then adding whatever custom headers we want before passing it off to client.Do.  As long as the underlying http.Client leaves those headers intact, then you're fine.  If your underlying http.Client is clobbering your auth header because it thinks it should be responsible for auth, then you do have a problem.  This is exactly why go-github's Client struct has its own Do and NewRequest methods, so that you can modify the request before calling Do.  Not separating those, as go-asana does with its single Request method, results in these weird workarounds.